### PR TITLE
API support and PriceList.

### PR DESCRIPTION
**Pull request by styxhope**
Create a global CommDKPApi that allows access to limited parts of CommDKP.
Introduce PriceList support through CommDKPApi:SetPriceListApi.

The PriceList api requires a method GetItemPrice(itemLink) which is required to return a { minBid = x, maxBid = y } or nil.

### CODE REVIEW

**Comment by Vapok** [comment]
Hi thanks for putting in a Pull Request. You'll have to do a little explaining here as to what you're doing with this. I am concerned about seeing that remove of the MaxBid functions, and some of the other changes that it looks like other parts of the addon rely on but I don't see those parts refactored.  Can you walk me through the changes here?

File: Modules/Bidding.lua [hunk 1]
```diff
@@ -339,84 +348,70 @@ function CommDKP:WhisperAvailableDKP(name, cmd)
   return;
 end
 
-function CommDKP:GetMinBid(itemLink)
```
Comment (Vapok) [inline Modules/Bidding.lua:342]: Why are you removing this function?

File: Modules/Bidding.lua [hunk 2]
```diff
@@ -339,84 +348,70 @@ function CommDKP:WhisperAvailableDKP(name, cmd)
...
   elseif loc == "INVTYPE_RANGED" or loc == "INVTYPE_THROWN" or loc == "INVTYPE_RANGEDRIGHT" or loc == "INVTYPE_RELIC" then
-    return core.DB.MinBidBySlot.Range
+    return from.Range
   else
-    return core.DB.MinBidBySlot.Other
+    return from.Other
   end
 end
 
-function CommDKP:GetMaxBid(itemLink)
```
Comment (Vapok) [inline Modules/Bidding.lua:382]: Why are you removing this function?

File: Modules/Bidding.lua (hunk 1 above)
Comment (Vapok) [reply Modules/Bidding.lua:342]: Okay, I see that you've added it below.  But I don't see a new GetMaxBid

File: Modules/Bidding.lua [hunk 3]
```diff
@@ -339,84 +348,70 @@ function CommDKP:WhisperAvailableDKP(name, cmd)
...
-  elseif loc == "INVTYPE_HOLDABLE" or loc == "INVTYPE_SHIELD" then
-    return core.DB.MaxBidBySlot.OffHand
-  elseif loc == "INVTYPE_RANGED" or loc == "INVTYPE_THROWN" or loc == "INVTYPE_RANGEDRIGHT" or loc == "INVTYPE_RELIC" then
-    return core.DB.MaxBidBySlot.Range
-  else
-    return core.DB.MaxBidBySlot.Other
+  return CommDKP:GetItemLocBid(core.DB.MinBidBySlot, itemLink)
+end
+
+function CommDKP:GetMaxBid(itemLink)
```
Comment (styxhope) [inline Modules/Bidding.lua:408]: You can see it added here again. It's just the diff tool.

Basically I just refactored the two methods to reduce the amount of code.

File: Modules/Bidding.lua (hunk 2 above)
Comment (styxhope) [reply Modules/Bidding.lua:382]: It is also still there, it's just the diff that says removed. You can see it added again on line 408.

**Comment by styxhope** [comment]
Basically the changes allow a thirdparty addon to call CommDKPApi:SetPriceListApi with a dictionary that has a GetItemPrice function.

So GetMinBid and GetMaxBid now calls into CommDKP:GetItemPrice. 
CommDKP:GetItemPrice now checks if such a callback has been registered, if so it calls that method to get a tuple { minBid = x, maxBid = y }. If such is found those values are used. If not it calls the default slot methods. 

The refactoring I did in GetMinBid and GetMaxBid where because they were identical, except from the variable the cost was pulled from. So I made one function which takes that variable as a parameter. Thus reducing the amount of duplicated code.

**Comment by styxhope** [comment]
The thirdparty Addon I'm writing is basically just a big table mapping a itemID to a minBid and a maxBid. But the infrastructure now allows anyone to build a complex or simple Addon to do their specific pricing if they want to. Without the need to modify CommunityDKP.

**Comment by styxhope** [comment]
Any chance of getting this reviewed?

**Review by Vapok — APPROVED** [review]

**Pull request merged**